KOGITO-766: Kogito Keyboard Shortcuts API

tiagobento opened the pull request:
On this PR:
- Using the new Keyboard Shortcuts API to register existing keyboard shortcuts on Kogito environment. Syntax is very simple and keep in mind that "ctrl" gets replaced with "cmd" when running on macOS. Should we need a more flexible API in the future, we can change it.

**VSIX for testing:** https://drive.google.com/file/d/1pKGSpn6xHm3sll8TH0XOc3fp6FMHjkN6/view?usp=sharing

Related PRs:
https://github.com/kiegroup/lienzo-core/pull/102
https://github.com/kiegroup/lienzo-tests/pull/84
https://github.com/kiegroup/appformer/pull/922
https://github.com/kiegroup/kie-wb-common/pull/3221
https://github.com/kiegroup/kogito-tooling/pull/86

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/canvas/controls/keyboard/shortcut/AppendDecisionShortcut.java:
```diff
@@ -68,4 +66,14 @@ private boolean selectedNodeIsInput(final Element selectedElement) {
     }
+
+    @Override
+    public String getKogitoCombination() {
```
I don't think we should litter _non_ `kogito` code with `kogito` terms.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/canvas/controls/keyboard/shortcut/AppendDecisionShortcut.java:
```diff
@@ -68,4 +66,14 @@ private boolean selectedNodeIsInput(final Element selectedElement) {
+    }
+
+    @Override
+    public String getKogitoLabel() {
```
I don't think we should litter _non_ `kogito` code with `kogito` terms.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/LocationControlImpl.java:
```diff
@@ -117,6 +120,29 @@ public LocationControlImpl(final CanvasCommandFactory<AbstractCanvasHandler> can
         // Keyboard event handling.
         session.getKeyboardControl().addKeyShortcutCallback(this::onKeyDownEvent);
+
+        //KOGITO
```
I don't think we should litter _non_ `kogito` code with `kogito` terms.

romartin replied:
agree - on the core modules there should be no reference to kogito (eg: `KogitoKeyPress`)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/pom.xml:
```diff
@@ -93,6 +93,11 @@
       <scope>provided</scope>
     </dependency>
 
+    <dependency>
```
I don't think we should litter _non_ `kogito` modules with dependencies on `kogito` modules. BPMN and DMN have two clear codebases; Business Central and `kogito`. Only the existing 'kogito` modules should have additional dependencies on `kogito` added.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyboardControl.java:
```diff
@@ -33,4 +34,122 @@
+
+    //
+    //
+    //Kogito
```
I don't think we should litter _non_ `kogito` code with `kogito` terms.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/shortcut/KeyboardShortcut.java:
```diff
@@ -40,4 +41,12 @@
     void executeAction(final H canvasHandler,
                        final String selectedNodeId);
+
+    String getKogitoCombination();
```
I don't think we should litter _non_ `kogito` code with `kogito` terms.

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyEventHandler.java:
```diff
@@ -52,9 +65,82 @@
 
     public KeyEventHandler addKeyShortcutCallback(final KeyboardControl.KeyShortcutCallback shortcutCallback) {
         this.shortcutCallbacks.add(shortcutCallback);
+
```
I don't think we should litter _non_ `kogito` code with `kogito` terms.

tiagobento commented:
Hi @manstis. Thanks for your review. I have some points I'd like to discuss before I start changing the code. Just to be clear, I agree that the more we can separate Kogito and non-Kogito code, the better. However, I don't think it will be the best choice for every case that you pointed out. See my observations:

1. The `KogitoKeyShortcutCallback` and its sub-classes are specializations of the existent `KeyShortcutCallback` and they provide methods that are only needed for Kogito, so I'm not sure how we could change that in order to separate things. They're used to call shortcuts registration with the necessary information and its use are always inside classes that represent/control "actions". For example, `LocationControlImpl`. I also can't see how we would change that to separate things.
2. The extra methods added to `KeyboardShortcut` were also aiming that. I can remove the "kogito" reference, but I thought I would make it explicit so that people know where it's used. The key combination and labels are intrinsic to each AppendNode action in this case, do you have any suggestions on how to address that?
3. I can try and create a different implementation of `KeyEventHandler` that's only used on kogito webapps. Is there a module I can place it already or should I create a new one?
4. The dependency added to `kie-wb-common-stunner-client-api/pom.xml` is used because we need the `Opts` class that's part of the new API. I also don't see how we can address that without having a duplicate of this class and making a mapping inside a "kogito" module..

Hope I could explain it properly! I'm open for ideas on how to make this integration better without sacrificing simplicity. Thanks!

tiagobento commented:
Jenkins, please retest this.

manstis commented:
@tiagobento OK, so I've had a better look at this PR and here's my thoughts:

Your change to `AbstractCanvasShortcutsControlImpl` appears to register individual `KeyboardShortcut` implementations to `KeyboardControlImpl` as _hybrid_ `KogitoKeyPress` versions of Stunner's `KeyShortcutCallback` (L#54). This is in addition to the existing binding that registers a single callback that handles all key presses using  `KeyboardShortcut.matchesPressedKeys` (L#52).

You've also changed `KeyEventHandler` (used by `KeyboardControlImpl` above where shortcuts have been registered) to handle both Stunner's existing mechanism and the integration with `appformer-kogito-bridge`. If the kogito environment is detected (checking for Errai's remote bus) you then delegate it to `KeyboardShortcutsApi` the implementation of which in `kogito-tooling` uses `KeyboardShortcut.getKogitoCombination()` to match key presses with key combinations. It uses both `KeyboardShortcut.getKogitoCombination()` and `KeyboardShortcut.getKogitoLabel()` to drive the content of the keyboard shortcuts popup.

So, IMO, given my [comment](https://github.com/kiegroup/appformer/pull/922#issuecomment-601393776) on Appformer I think it might be advisable to consider:
- Remove `getKogitoCombination()` from `KeyboardShortcut`
- Rename `getKogitoLabel()` in `KeyboardShortcut` to something like `getDescription()`
- Rename `getKogitoOpts() in `KeyboardShortcut` to something like `getOptions()`
- Add `getKeys()` to `KeyboardShortcut` that returns a `List` of `KeyboardEvent.Key` representing the key combinations to which the shortcut responds. The return value from this can then be used directly in the various `KeyboardShortcut.matchesPressedKeys(..)` implementations.
- Change `AbstractCanvasShortcutsControlImpl` to register individual `KeyboardShortcut` to something like a `KeyboardShortcutPublisher` (in `appformer-client-api`) instead of to `KeyboardControlImpl`
- Write a default `KeyboardShortcutPublisher` in `appformer-client-api` that has a NOP function in Business Central.
- Write a `kogito` version of `KeyboardShortcutPublisher` in `appformer-kogito-bridge` with associated producer for use in kogito channels.
- The `kogito` version of `KeyboardShortcutPublisher` does in essence what your changes to `KeyEventHandler` does. It would however map `KeyboardShortcut.getKeys()` to the string used by `KeyboardShortcutsApi` (representing the current `getKogitoCombination()` method). 
- Your changes to `KeyboardControl` for `kogito` are all moved to `KeyboardShortcutPublisher` (or at least somewhere in `appformer-client-api`) as they're only needed for the bridge between the kogito implementation of `KeyboardShortcutPublisher` and `KeyboardShortcutsApi`


I believe this would provide a cleaner separation of concerns for non-kogito and kogito environment implementations of the keyboard shortcut _publication_ mechanism. I think lack of a really clean separation of definitions and implementation of _features_ in Appformer has led to the difficulty we've had trying to minimise the dependencies in the editor base WARs. Let's not add more difficulties for future separation.

Anyway.. this is purely _my_ opinion of probably several people who should look at this PR.

My main issue with it as it stands are:
- Mention of `kogito` in non-kogito related modules
- Necessity for `kogtio` dependencies in non-kogito related modules.
- Duplication of shortcut declarations (`matchesPressedKeys(..) implementations and `getKogitoCombination()`)

This has turned into a bit of a dissertation. Sorry.

tiagobento commented:
HI @manstis! Thx for taking the time to carefully look at this new feature. I have some counter-points to what you listed:

1)
- Remove getKogitoCombination() from KeyboardShortcut
- Rename getKogitoLabel() in KeyboardShortcut to something like getDescription()
- Rename getKogitoOpts() in KeyboardShortcut to something like getOptions()`
- Add getKeys() to KeyboardShortcut that returns a List of KeyboardEvent.Key representing the key combinations to which the shortcut responds. The return value from this can then be used directly in the various KeyboardShortcut.matchesPressedKeys(..) implementations.
> These points make total sense when you think about reusing logic and requirements for both BC and Kogito. However, when developing this feature, I thought that we should, in fact, separate things. I don't want BC using these KogitoLabels nor do I want BC using the exact same key combinations. I designed that in a way that we can have different key bindings for the same action in Kogito and BC. IMHO, Kogito is moving in a very fast pace and we're in a very complex environment (given the many channels), so in _my opinion_ we should not "merge" both BC and Kogito shortcuts logic.
Also, your suggestion does not solve the dependency on `appformer-kogito-bridge`, since that's where the Opts object comes from. To solve that, the only way I can think of is writing a map on a kogito-specific module where we create `Opts` objects for each KeyboardShortcut implementation. Don't know how you feel about that though. This still maintains the property of having potentially different key combinations for Kogito and BC, which I like.

2)
- Change AbstractCanvasShortcutsControlImpl to register individual KeyboardShortcut to something like a KeyboardShortcutPublisher (in appformer-client-api) instead of to KeyboardControlImpl
- Write a default KeyboardShortcutPublisher in appformer-client-api that has a NOP function in Business Central.
- Write a kogito version of KeyboardShortcutPublisher in appformer-kogito-bridge with associated producer for use in kogito channels.
> Agree with that. I can create a new alternative that knows what to do with all the `KeyboardShortcut` instances on the Kogito case. The only thing is that it can't be on `appformer`, since all the interfaces and models are defined directly on `kie-wb-common`. So I think we'll need a new `kie-wb-common-kogito-client` module or something like that?

3)
- The kogito version of KeyboardShortcutPublisher does in essence what your changes to KeyEventHandler does. It would however map KeyboardShortcut.getKeys() to the string used by KeyboardShortcutsApi (representing the current getKogitoCombination() method).
- Your changes to KeyboardControl for kogito are all moved to KeyboardShortcutPublisher (or at least somewhere in appformer-client-api) as they're only needed for the bridge between the kogito implementation of KeyboardShortcutPublisher and KeyboardShortcutsApi
> There are two ways to register keyboard shortcuts at the moment. Implement `KeyboardShortcut` class or call `KeyboardControl` directly. On the first case, we would go through the alternative described in 2). On the second one, I think our best solution is to create an alternative to `KeyEventHandler`. This alternative would be what you described as the `KeyboardShortcutPublisher`. This would also live in this new module `kie-wb-common-kogito-client`.

4)
> You didn't mention that directly, so I'm bringing that up. Consider the case on `LienzoCanvasMediators`. We're using the editor context to decide whether to use CMD or CTRL to zoom. Which we should address as well. Probably with an alternative or something like that, but also, we're calling `KeyboardControl` directly, using Kogito-specific interfaces. These interfaces are completely different from the non-Kogito ones. They end up in `KeyEventHandler`, where we sort them out (btw, just fixed an issue there, see db3d93ebfd1879825903ad34718885b534ed2125). Without creating a "generic" API, I think it'll be very hard to remove those calls. We probably should have them stay there? Idk, this is all too complicated :) Maybe we should just accept that and mark somewhere that we have this technical debt and revisit it in case we separate Kogito and non-Kogito?

So now about your main concerns, I think we won't solve them, since the `Opts` object itself makes a non-Kogito module have a dependency to a Kogito one. That's really the only hard dependency. On a separation scenario, we would have to simply remove those methods from the BC version and remove the dependency to `appformer-kogito-bridge`. However, if we go with a map mentioned in 1), we can avoid all that and have the separation through the expense of remembering to add a new map entry every time a new `KeyboardShortcut` is created.

To summarise, my proposal is:
**a.** Create a new module on kie-wb-common where kogito-specific logic will live. Like a `kie-wb-common-client-kogito`. Please tell me if you have a better name 😅 
**b.** Leave kogito key combinations separate from BC's and have the extra methods on a non-Kogito interface, knowingly adding a dependency OR create the mapping I mentioned.
**c.** Create an alternative to the `AbstractCanvasShortcutsControlImpl` class. Probably here's where said `KeyboardShortcut` mapping would live if we go with that.
**d.** Create an alternative to the `KeyEventHandler`, where the `KeyboardShortcutsApi` from `appformer-kogito-bridge` would be used.
**e.** Talk about 4) better.

Thx again for reading all that! Let me know what you think :)

manstis commented:
I feel we're in danger of repeating the UK's Brexit negotiations...

> 1., 2., 3
- Quick win: [kie-wb-common-kogito-client](https://github.com/kiegroup/kie-wb-common/tree/master/kie-wb-common-kogito/kie-wb-common-kogito-client).... HOWEVER that cannot depend on Stunner modules (as Stunner's `kogtio` related modules depend on this) SO [kie-wb-common-stunner-kogito-client](https://github.com/kiegroup/kie-wb-common/tree/master/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-kogito/kie-wb-common-stunner-kogito-client) is probably a better home.

- IF you want to support different keyboard bindings for Business Central vs kogito (what a crazy world that would be for Users... anyway...) then IMO different `@Alternative` implementations of the `KeyboardShortcut` classes should be added to `kie-wb-common-kogito-client` would be the way to go. Have you ran the notion of different keyboard bindings in BC vs kogito past UX?!

- `Opts` could be moved to `appformer-client-api` (`appformer-kogito-bridge` already depends on it).

- I agree about an `@Alternative` for `KeyEventHandler` for `kogito` (serving as the _publisher_).

- Add `getLabel()` to `KeyboardShortcut` to suit your "Keyboard shortcuts" popup in `kogito`.

These should allow us to kill all reference to `kogito` in non-`kogito` modules!

> 4.
- If you're calling `KeyboardControl` with `kogito` specific interfaces (I guess you mean those defined [here](https://github.com/kiegroup/kie-wb-common/pull/3221/files#diff-e91b688e5c88e61ec1a5078062f7d1c3R40)) then provided their sub-interfaces (defined in `kie-wb-common-stunner-kogito-client`) they'll end up in the `@Alternative` `KeyEventHandler` and can be handled accordingly (publishing them to `KeyboardShortcutsApi`
- Or we need to talk about 4) better :-)

romartin commented:
Hopefullly I'll take a look on Monday guys, but in general I'm agree with @manstis comments, we should not add more coupling on kogito and non-kogito things, so lemme know once I can start the review again.

tiagobento commented:
@manstis @romartin Hi! First of all thank you so much for your feedback. I'm working on making this change decoupled with Kogito now. As you might expect, I'm creating some `@Alternatives`. Other than the two Kogito DMN and BPMN apps, do you know where should I add the non-Kogito alternatives to?

tiagobento commented:
Things I already did:
1. Used KeyboardEvent.Keys to register keybindings.
2. Created alternative to register `KeyboardShortcuts`
3. Created alternative to `KeyEventHandler`
4. Moved EditorContext to `appformer-client-api` so that we can use it in non-Kogito modules as well.
5. Moved KeyboardShortcutsApiOpts to `appformer-client-api` so that we can use it in non-Kogito modules as well.

Things I still have to do:
1. Remove direct imperative calls to KeyboardControlImpl
2. Configure alternatives for BC compilation

Please tell me if I'm forgetting something.

manstis commented:
@romartin @jomarko, @ederign wants this in DevPreview but @tiagobento may, AFAIU, not be able to complete the refactoring I've requested to separate `kogito-tooling` needs from non`-kogito` Stunner modules. Personally I don't think we should merge w/o the correct separation (as the problems we had making a minimal editor in the first place was due to spaghetti dependencies).

However, given I am moving to TrustyAI and my opinion is only one of several needed to review these changes. Please take a look; if you're OK with it then please; I will not block merge.

There would be additional changes needed after merge to complete separation. It's been proposed somebody within the DMN, BPMN or Stunner teams would be able to complete this work as the `kogito-tooling` people are too busy....

manstis commented:
@karreiro Sorry, I've added you too (probably as an optional reviewer).

With my scheduled move it's likely you'll be involved with work around this area at some point.

tiagobento commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyEventHandlerImpl.java:
```diff
@@ -0,0 +1,137 @@
+ * multiple key events.
+ */
+@Dependent
+@Alternative
```
Yes, exactly. That was my question on https://github.com/kiegroup/kie-wb-common/pull/3221#issuecomment-602961421. This one is more a default implementation than an Alternative, but I couldn't figure out a way to do that. Can you point me to somewhere where something like that is done? It'd be really helpful.

tiagobento replied:
@manstis ^

manstis replied:
@tiagobento OK, then you can simply remove the `@Alternative` annotation and Java's CDI will discover this class and use it as the _default_ implementation.

tiagobento replied:
Uau, really? That's nice (and easy 🤩). So just to be clear, we have 2 implementations --. Kogito and non-Kogito. If only the Kogito implementation has the @Alternative annotation, then on non-Kogito webapps, there will only be one implementation, which is fine and will work. But on Kogito webapps I have to enable the alternative. This will will make CDI ignore the non-Kogito implementation and inject the enabled-via-alternatives Kogito implementation!?

manstis replied:
Correct.

tiagobento replied:
@manstis Hi! So I did a test like you suggested and removed the `@Alternative` from the non-Kogito implementations. It compiles okay, but on Kogito editors, there's a runtime error saying that there are ambiguous dependency resolutions. Am I missing something?

manstis replied:
Hi @tiagobento following your comment above; I cloned these PRs again and deleted `@Alternative` from `org.kie.workbench.common.stunner.core.client.canvas.controls.keyboard.KeyEventHandlerImpl` (as this is the _default_ implementation in Stunner). I see you've enabled the alternative [here](https://github.com/kiegroup/kie-wb-common/pull/3221/files#diff-46fbcfb569443805a1932efaaef49f40R37) and [here](https://github.com/kiegroup/kie-wb-common/pull/3221/files#diff-e62febcaf861692660f22ccaba5e4421R36) so did not need to make any other local changes.

 I then compiled it all again and launched `kie-wb-common-stunner-bpmn-kogito-runtime` with `mvn clean gwt:run`. It worked fine. No errors relating to ambiguous dependencies. I also tested `kie-wb-common-dmn-webapp-kogito-runtime` and that was OK too. 

Perhaps you was lucky enough to have local/environmental/cache issues?

tiagobento commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyboardControl.java:
```diff
@@ -33,4 +34,122 @@
+        String getLabel();
+    }
+
+    class KogitoKeyPress implements KogitoKeyShortcutCallback {
```
See comment below.

tiagobento commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyboardControl.java:
```diff
@@ -33,4 +34,122 @@
+    //
+    //Kogito
+
+    interface KogitoKeyShortcutCallback extends KeyShortcutCallback {
```
See comment below.

tiagobento commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyboardControl.java:
```diff
@@ -33,4 +34,122 @@
+        }
+    }
+
+    class KogitoKeyShortcutKeyDownThenUp implements KogitoKeyShortcutCallback {
```
See comment below.See comment below.

tiagobento commented:
Hi @manstis. Thanks for revisiting this PR! I'm glad you like the changes. I also think these changes were a great improvement over what we had before.

About the missing bits, like you already know, it's more complicated. As we discussed before, my proposal is to postpone that for after Dev Preview. Ideally, someone with more knowledge on the subject would do this separation.

In a scenario where my suggestion is accepted, we can also talk about the names of these Kogito-specific interfaces. I'm okay to change them if you think having a "Kogito" reference in non-Kogito modules is worse. IMHO, it's better to leave them as they are, since whoever works on the separation (myself included) will have a better pointer for where to look. WDYT?

Thx!

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/canvas/controls/keyboard/shortcut/AppendDecisionShortcut.java:
```diff
@@ -68,4 +67,14 @@ private boolean selectedNodeIsInput(final Element selectedElement) {
+
+    @Override
+    public String getLabel() {
+        return "Append Decision";
```
missing i18n

tiagobento replied:
@romartin Can you point me to a class that resolves i18n keys that could be used here?

tiagobento replied:
Btw this string will only appear on the Kogito editors.

tiagobento commented:
@romartin @manstis @karreiro I'm currently working on other tasks right now since priorities changed a bit. There is still work to do on this PR and on https://github.com/kiegroup/kogito-tooling/pull/86. I'll let you know when I resume this task.

Remaining work:
1. Apply changes done in 7390044 to bpmn as well. (Default impls)
2. Ignore Kogito-related interfaces on `KeyEventHandlerImpl#addKeyShortcutCallback`
3. Review API and the need of exposing event's `target` attribute.

romartin commented:
Hey @tiagobento 

ok thanks for the feedback. We're having now more "free" time, so once going back to this lemme know and I'll help as much as possible too, maybe we can even do some call to clarify the implementation or scope for this as well, np for me! But yeah, let's now try to move this fordward, just lemme know once back!

Thanks!

tiagobento commented:
@romartin Thx! I'll reach out as I resume this task! :)

tiagobento commented:
@romartin Hi Roger! So I've resumed this task (yay! 🎉) and since your offered help a while ago I figured we could talk about the missing item on https://github.com/kiegroup/kie-wb-common/pull/3221#issuecomment-610384482.

Michael has found a way to properly de-couple Kogito and non-Kogito dependencies, so I'd like to know your thoughts on that (that's the commit -> 7390044). Michael has changed things for DMN, but we need the sam for BPMN. Also, since SceSim has an editor now, we'll probably do the same there.

Let me know if you're available to talk about that! We can schedule a call to if you'd like :)

tiagobento commented:
@karreiro As requested, `KeyboardEvent.target` has been exposed on the API. Please re-review that once you can :)

tiagobento commented:
@romartin @karreiro This PR is ready for review. Thanks everyone for the help so far! :)

tiagobento commented:
Btw, i18n will be addressed on this JIRA: https://issues.redhat.com/browse/KOGITO-2120

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this.

tiagobento commented:
@romartin Can you please review that?

ederign commented:
@jomarko @romartin can you please take a look at this? I really would like to have this merged before the end of the sprint (if everything is ok)

romartin commented:
@tiagobento I see you added some commit since last time we talked on this. I've also been noticed on some use cases to test for the keyboards stuff. If any issues appear, althought maybe not related to this work, as those are already present, I'll just report them and do not prevent this for being merged, but would really like to give a last try on this before finally approving, at least just to figure out the final state for this. So can you maybe please send me again, offline, the vsix for testing this PR on the latest state please? just send me the link offline and I'll proceed with testing, it'll be just a minute, so hopefully we're close to merge this anyway.

Thx!

jomarko commented:
Getting this error when building `kie-wb-distributions/business-central-parent/business-central-webapp` locally as `mvn clean install` 
```
AuthorityRequirement.title, org.kie.workbench.common.stunner.bpmn.definition.property.task.MultipleInstanceExecutionMode.description]
[INFO] 10:29:52.368 [pool-4-thread-3] INFO  o.j.e.m.r.MarshallerGeneratorFactory - generated marshaller factory class in 15114ms.
[INFO] 10:30:00.296 [pool-4-thread-3] INFO  com.google.gwt.core.ext.Generator - ** deposited marshaller class in : /home/jomarko/redhat/github/kiegroup/kie-wb-distributions/business-central-parent/business-central-webapp/target/business-central-webapp/WEB-INF/classes
[INFO] 10:30:00.299 [pool-4-thread-3] INFO  o.j.e.m.r.MarshallerGeneratorFactory - generating marshaller factory class for client...
[INFO] 10:30:00.471 [pool-4-thread-3] INFO  o.j.e.m.r.MarshallerGeneratorFactory - generated marshaller factory class in 172ms.
[INFO]       Rebinding org.jboss.errai.marshalling.client.api.MarshallerFactory
[INFO]          Invoking generator org.jboss.errai.marshalling.rebind.MarshallersGenerator
[INFO]             Generating Marshallers Bootstrapper...
[INFO] 10:30:03.087 [pool-4-thread-7] INFO  o.j.e.i.r.i.b.IOCProcessor - Reachability strategy set to Annotated
[INFO]    Computing all possible rebind results for 'org.jboss.errai.ioc.client.container.IOCEnvironment'
[INFO]       Rebinding org.jboss.errai.ioc.client.container.IOCEnvironment
[INFO]          Invoking generator org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCEnvironmentGenerator
[INFO]             Generating Extensions Bootstrapper...
[INFO]       Rebinding org.jboss.errai.ioc.client.container.IOCEnvironment
[INFO]          Invoking generator org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCEnvironmentGenerator
[INFO]             Generating Extensions Bootstrapper...
[INFO]    Computing all possible rebind results for 'org.jboss.errai.ioc.client.Bootstrapper'
[INFO]       Rebinding org.jboss.errai.ioc.client.Bootstrapper
[ERROR] java.util.concurrent.ExecutionException: java.lang.RuntimeException: The following dependency problems were found:
[INFO]          Invoking generator org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator
[INFO]             generating ioc bootstrapping code...
[ERROR] 	Unsatisfied constructor dependency @Default org.appformer.client.context.EditorContextProvider for Class @Any @Default org.kie.workbench.common.stunner.client.lienzo.components.mediators.LienzoCanvasMediators. Some beans were found that satisfied this dependency, but must be enabled:
[INFO]             [ERROR] Error generating BootstrapperImpl
[ERROR] org.appformer.client.context.DefaultEditorContextProviderImpl
[INFO] java.util.concurrent.ExecutionException: java.lang.RuntimeException: The following dependency problems were found:
[INFO] 	Unsatisfied constructor dependency @Default org.appformer.client.context.EditorContextProvider for Class @Any @Default org.kie.workbench.common.stunner.client.lienzo.components.mediators.LienzoCanvasMediators. Some beans were found that satisfied this dependency, but must be enabled:
[ERROR] 
[INFO] org.appformer.client.context.DefaultEditorContextProviderImpl
[ERROR] 
[INFO] 
[ERROR] 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
[INFO] 
[ERROR] 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
[INFO] 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
[ERROR] 	at org.jboss.errai.config.rebind.AsyncGenerators$FutureWrapper.get(AsyncGenerators.java:112)
[ERROR] 	at org.jboss.errai.config.rebind.AsyncGenerators$FutureWrapper.get(AsyncGenerators.java:86)
[INFO] 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
[ERROR] 	at org.jboss.errai.config.rebind.AbstractAsyncGenerator.startAsyncGeneratorsAndWaitFor(AbstractAsyncGenerator.java:205)
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator.generate(IOCGenerator.java:62)
[INFO] 	at org.jboss.errai.config.rebind.AsyncGenerators$FutureWrapper.get(AsyncGenerators.java:112)
[ERROR] 	at com.google.gwt.core.ext.IncrementalGenerator.generateNonIncrementally(IncrementalGenerator.java:40)
[INFO] 	at org.jboss.errai.config.rebind.AsyncGenerators$FutureWrapper.get(AsyncGenerators.java:86)
[ERROR] 	at com.google.gwt.dev.javac.StandardGeneratorContext.runGeneratorIncrementally(StandardGeneratorContext.java:745)
[INFO] 	at org.jboss.errai.config.rebind.AbstractAsyncGenerator.startAsyncGeneratorsAndWaitFor(AbstractAsyncGenerator.java:205)
[ERROR] 	at com.google.gwt.dev.cfg.RuleGenerateWith.realize(RuleGenerateWith.java:103)
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator.generate(IOCGenerator.java:62)
[INFO] 	at com.google.gwt.core.ext.IncrementalGenerator.generateNonIncrementally(IncrementalGenerator.java:40)
[INFO] 	at com.google.gwt.dev.javac.StandardGeneratorContext.runGeneratorIncrementally(StandardGeneratorContext.java:745)
[INFO] 	at com.google.gwt.dev.cfg.RuleGenerateWith.realize(RuleGenerateWith.java:103)
[INFO] 	at com.google.gwt.dev.shell.StandardRebindOracle$Rebinder.rebind(StandardRebindOracle.java:78)
[INFO] 	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:262)
[INFO] 	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:251)
[INFO] 	at com.google.gwt.dev.PrecompilationContextCreator$1.getAllPossibleRebindAnswers(PrecompilationContextCreator.java:86)
[ERROR] 	at com.google.gwt.dev.shell.StandardRebindOracle$Rebinder.rebind(StandardRebindOracle.java:78)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createStaticRebindExpression(UnifyAst.java:519)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createRebindExpression(UnifyAst.java:487)
[ERROR] 	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:262)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.maybeHandleMagicMethodCall(UnifyAst.java:415)
[ERROR] 	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:251)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.visit(UnifyAst.java:402)
[INFO] 	at com.google.gwt.dev.jjs.ast.JMethodCall.traverse(JMethodCall.java:265)
[ERROR] 	at com.google.gwt.dev.PrecompilationContextCreator$1.getAllPossibleRebindAnswers(PrecompilationContextCreator.java:86)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[ERROR] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createStaticRebindExpression(UnifyAst.java:519)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[ERROR] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createRebindExpression(UnifyAst.java:487)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[INFO] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
[INFO] 	at com.google.gwt.dev.jjs.ast.JCastOperation.traverse(JCastOperation.java:76)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[INFO] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
[INFO] 	at com.google.gwt.dev.jjs.ast.JDeclarationStatement.traverse(JDeclarationStatement.java:49)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
[INFO] 	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
[ERROR] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.maybeHandleMagicMethodCall(UnifyAst.java:415)
[INFO] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
[INFO] 	at com.google.gwt.dev.jjs.ast.JTryStatement.traverse(JTryStatement.java:93)
[ERROR] 	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.visit(UnifyAst.java:402)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JMethodCall.traverse(JMethodCall.java:265)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JCastOperation.traverse(JCastOperation.java:76)
[INFO] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at com.google.gwt.dev.jjs.ast.JMethodBody.traverse(JMethodBody.java:83)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JDeclarationStatement.traverse(JDeclarationStatement.java:49)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
[INFO] 	at com.google.gwt.dev.jjs.ast.JMethod.visitChildren(JMethod.java:786)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
[INFO] 	at com.google.gwt.dev.jjs.ast.JMethod.traverse(JMethod.java:778)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.mainLoop(UnifyAst.java:1401)
[INFO] 	at com.google.gwt.dev.jjs.impl.UnifyAst.exec(UnifyAst.java:896)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.unifyJavaAst(JavaToJavaScriptCompiler.java:1410)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.constructJavaAst(JavaToJavaScriptCompiler.java:1222)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:1140)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JTryStatement.traverse(JTryStatement.java:93)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
[INFO] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:255)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
[INFO] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:255)
[INFO] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:202)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:143)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:204)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
[INFO] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:155)
[INFO] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:144)
[INFO] 	at com.google.gwt.dev.Compiler$1.run(Compiler.java:118)
[INFO] 	at com.google.gwt.dev.CompileTaskRunner.doRun(CompileTaskRunner.java:55)
[INFO] 	at com.google.gwt.dev.CompileTaskRunner.runWithAppropriateLogger(CompileTaskRunner.java:50)
[INFO] 	at com.google.gwt.dev.Compiler.main(Compiler.java:125)
[INFO] Caused by: java.lang.RuntimeException: The following dependency problems were found:
[INFO] 	Unsatisfied constructor dependency @Default org.appformer.client.context.EditorContextProvider for Class @Any @Default org.kie.workbench.common.stunner.client.lienzo.components.mediators.LienzoCanvasMediators. Some beans were found that satisfied this dependency, but must be enabled:
[ERROR] 	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
[INFO] org.appformer.client.context.DefaultEditorContextProviderImpl
[ERROR] 	at com.google.gwt.dev.jjs.ast.JMethodBody.traverse(JMethodBody.java:83)
[INFO] 
[INFO] 
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.graph.impl.DependencyGraphBuilderImpl.resolveDependencies(DependencyGraphBuilderImpl.java:393)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.graph.impl.DependencyGraphBuilderImpl.createGraph(DependencyGraphBuilderImpl.java:172)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCProcessor.process(IOCProcessor.java:201)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCBootstrapGenerator.generateBootstrappingClassSource(IOCBootstrapGenerator.java:237)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JMethod.visitChildren(JMethod.java:786)
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCBootstrapGenerator.generate(IOCBootstrapGenerator.java:132)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JMethod.traverse(JMethod.java:778)
[INFO] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator.generate(IOCGenerator.java:72)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
[INFO] 	at org.jboss.errai.config.rebind.AbstractAsyncGenerator$1.call(AbstractAsyncGenerator.java:66)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
[INFO] 	at org.jboss.errai.config.rebind.AbstractAsyncGenerator$1.call(AbstractAsyncGenerator.java:54)
[ERROR] 	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
[INFO] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ERROR] 	at com.google.gwt.dev.jjs.impl.UnifyAst.mainLoop(UnifyAst.java:1401)
[INFO] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[INFO] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[INFO] 	at java.lang.Thread.run(Thread.java:748)
[INFO]       Rebinding org.jboss.errai.ioc.client.Bootstrapper
[INFO]          Invoking generator org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator
[ERROR] 	at com.google.gwt.dev.jjs.impl.UnifyAst.exec(UnifyAst.java:896)
[INFO]             generating ioc bootstrapping code...
[ERROR] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.unifyJavaAst(JavaToJavaScriptCompiler.java:1410)
[INFO]    [WARN] For the following type(s), generated source was never committed (did you forget to call commit()?)
[ERROR] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.constructJavaAst(JavaToJavaScriptCompiler.java:1222)
[INFO]       [WARN] org.jboss.errai.ioc.client.BootstrapperImpl
[ERROR] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:1140)
[INFO]    [ERROR] Could not find org.jboss.errai.ioc.client.BootstrapperImpl in types compiled from source. Is the source glob too strict?
[INFO]    [ERROR] Errors in 'org/jboss/errai/ioc/client/Container.java'
[ERROR] 	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.precompile(JavaToJavaScriptCompiler.java:255)
[INFO]       [ERROR] Line 70: Rebind result 'org.jboss.errai.ioc.client.BootstrapperImpl' could not be found
[ERROR] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:255)
[ERROR] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:202)
[ERROR] 	at com.google.gwt.dev.Precompile.precompile(Precompile.java:143)
[ERROR] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:204)
[ERROR] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:155)
[ERROR] 	at com.google.gwt.dev.Compiler.compile(Compiler.java:144)
[ERROR] 	at com.google.gwt.dev.Compiler$1.run(Compiler.java:118)
[ERROR] 	at com.google.gwt.dev.CompileTaskRunner.doRun(CompileTaskRunner.java:55)
[ERROR] 	at com.google.gwt.dev.CompileTaskRunner.runWithAppropriateLogger(CompileTaskRunner.java:50)
[ERROR] 	at com.google.gwt.dev.Compiler.main(Compiler.java:125)
[ERROR] Caused by: java.lang.RuntimeException: The following dependency problems were found:
[ERROR] 	Unsatisfied constructor dependency @Default org.appformer.client.context.EditorContextProvider for Class @Any @Default org.kie.workbench.common.stunner.client.lienzo.components.mediators.LienzoCanvasMediators. Some beans were found that satisfied this dependency, but must be enabled:
[ERROR] org.appformer.client.context.DefaultEditorContextProviderImpl
[ERROR] 
[ERROR] 
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.graph.impl.DependencyGraphBuilderImpl.resolveDependencies(DependencyGraphBuilderImpl.java:393)
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.graph.impl.DependencyGraphBuilderImpl.createGraph(DependencyGraphBuilderImpl.java:172)
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCProcessor.process(IOCProcessor.java:201)
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCBootstrapGenerator.generateBootstrappingClassSource(IOCBootstrapGenerator.java:237)
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCBootstrapGenerator.generate(IOCBootstrapGenerator.java:132)
[ERROR] 	at org.jboss.errai.ioc.rebind.ioc.bootstrapper.IOCGenerator.generate(IOCGenerator.java:72)
[ERROR] 	at org.jboss.errai.config.rebind.AbstractAsyncGenerator$1.call(AbstractAsyncGenerator.java:66)
[ERROR] 	at org.jboss.errai.config.rebind.AbstractAsyncGenerator$1.call(AbstractAsyncGenerator.java:54)
[ERROR] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ERROR] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[ERROR] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[ERROR] 	at java.lang.Thread.run(Thread.java:748)
[INFO]    Computing all possible rebind results for 'com.allen_sauer.gwt.dnd.client.util.DragClientBundle'
[INFO]       Rebinding com.allen_sauer.gwt.dnd.client.util.DragClientBundle
[INFO]          Invoking generator com.google.gwt.resources.rebind.context.InlineClientBundleGenerator
[INFO]             Preparing method css
[INFO]                The following problems were detected
[INFO]                   [WARN] Line 63 column 30: encountered "=". Was expecting one of: "+" "-" "," "/" ")" <STRING> <IDENT> <NUMBER> <URL> <PERCENTAGE> <PT> <MM> <CM> <PC> <IN> <PX> <EMS> <EXS> <DEG> <RAD> <GRAD> <MS> <SECOND> <HZ> <KHZ> <DIMEN> <HASH> <UNICODERANGE> <FUNCTION> 
[INFO] 10:30:05.784 [main] INFO  o.j.e.i.r.i.b.QualifierEqualityFactoryGenerator - Generating QualifierEqualityFactory...
[INFO] 10:30:05.809 [main] INFO  o.j.e.i.r.i.b.QualifierEqualityFactoryGenerator - Generated QualifierEqualityFactory in 25ms
```

jomarko commented:
jenkins execute full downstream build

tiagobento commented:
@jomarko I'll investigate the error of this comment. https://github.com/kiegroup/kie-wb-common/pull/3221#issuecomment-631960529

tiagobento commented on kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-runtime/src/test/java/org/kie/workbench/common/dmn/showcase/client/selenium/DMNDesignerKogitoSeleniumIT.java:
```diff
@@ -238,43 +237,7 @@ public void testBusinessKnowledgeModel() throws Exception {
                 .areIdentical();
     }
-
-    @Test
-    public void testCopyAndPaste() throws Exception {
-        final String source = loadResource("business-knowledge-model.xml");
-        setContent(source);
-
-        expandDecisionNavigatorDock();
-
-        final String nodeName = "BusinessKnowledgeModel-1";
-        final String copiedNodeName = "BusinessKnowledgeModel-2";
-        waitOperation()
-                .withMessage(format(NOT_PRESENT_IN_NAVIGATOR, nodeName))
-                .until(element(NODE, nodeName))
-                .click();
-
-        final WebElement canvas = waitOperation()
-                .withMessage(format(NOT_PRESENT_IN_NAVIGATOR, "canvas"))
-                .until(presenceOfElementLocated(xpath(CANVAS)));
-
-        final Actions actions = new Actions(driver);
-        actions.keyDown(Keys.CONTROL).sendKeys("c").keyUp(Keys.CONTROL).perform();
-        actions.keyDown(canvas, Keys.CONTROL).sendKeys("v").keyUp(Keys.CONTROL).perform();
-
-        collapseDecisionNavigatorDock();
-
-        final String actual = getContent();
-        XmlAssert.assertThat(actual)
-                .withNamespaceContext(NAMESPACES)
-                .valueByXPath("count(//dmn:businessKnowledgeModel[@name='" + nodeName + "'])")
-                .isEqualTo(1);
-
-        XmlAssert.assertThat(actual)
-                .withNamespaceContext(NAMESPACES)
-                .valueByXPath("count(//dmn:businessKnowledgeModel[@name='" + copiedNodeName + "'])")
-                .isEqualTo(1);
-    }
```
Copy/paste has moved from from the editors, to kogito-tooling. That means that it's not possible to use these shortcuts on this editors in isolation, since they're built targeting Kogito. To ensure coverage, we'll have a manual test case in our release sanity checks for now.

tiagobento commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/session/DMNCanvasShortcutsControl.java:
```diff
@@ -0,0 +1,27 @@
+import org.kie.workbench.common.stunner.core.client.canvas.controls.keyboard.KeyboardControl;
+import org.kie.workbench.common.stunner.core.client.session.impl.EditorSession;
+
+public interface DMNCanvasShortcutsControl extends CanvasControl<AbstractCanvasHandler>,
```
Sure.

tiagobento commented:
@jomarko For some reason I can't comment on https://github.com/kiegroup/kie-wb-common/pull/3221#discussion_r428486224, but i18n will be addressed on a separate JIRA [1].

[1] https://issues.redhat.com/browse/KOGITO-2120

tiagobento commented:
@jomarko As of mixing Kogito and non-Kogito modules, we only postponed one point, that is the `KeyboardControl.java`.

tiagobento commented:
jenkins execute full downstream build

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-client-api/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyboardControl.java:
```diff
@@ -33,4 +34,122 @@
+
+        KeyboardEvent.Key[] getKeyCombination();
+
+        String getLabel();
```
I think mentioned also in kogito-tooling part. I see crucial to document somewhere the needed format of label , which is "Category of shortcut | Action description". Even more clear would be split label into two parameters.

tiagobento replied:
Added Javadoc to this method and to the methods on the KeyboardShortcutsApi as well.

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/keyboard/KeyEventHandlerImpl.java:
```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
2020

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

jomarko commented:
No more comments from me. Thank you.

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
Jenkins please retest this

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build

tiagobento commented:
jenkins execute full downstream build
